### PR TITLE
Crossbeam (channels only)

### PR DESCRIPTION
**Pull request by entropin**
First, change is needed for allowing for multiple video receivers.

### CODE REVIEW

File: crates/core/Cargo.toml [hunk 1]
```diff
@@ -7,6 +7,7 @@ edition = "2018"
 
 [dependencies]
 aes = "0.6"
+crossbeam = "0.7"
```
Comment (QuantumEntangledAndy) [inline crates/core/Cargo.toml:10]: I think you only need the channel

```suggestion
crossbeam_channel = "0.7"
```

File: crates/core/src/bc_protocol/binarysub.rs [hunk 2]
```diff
@@ -1,9 +1,7 @@
 use super::{connection::BcSubscription, RX_TIMEOUT};
 use crate::bc::model::*;
-use std::{
-    io::{BufRead, Error, ErrorKind, Read},
-    sync::mpsc::RecvTimeoutError,
-};
+use crossbeam::RecvTimeoutError;
```
Comment (QuantumEntangledAndy) [inline crates/core/src/bc_protocol/binarysub.rs:3]:
```suggestion
use crossbeam_channel::RecvTimeoutError;
```

File: crates/core/src/bc_protocol/connection.rs [hunk 3]
```diff
@@ -1,13 +1,13 @@
 use crate::bc;
 use crate::bc::model::*;
+use crossbeam::{unbounded, Receiver, Sender};
```
Comment (QuantumEntangledAndy) [inline crates/core/src/bc_protocol/connection.rs:3]:
```suggestion
use crossbeam_channel::{unbounded, Receiver, Sender};
```

File: crates/core/src/bc_protocol/connection.rs [hunk 4]
```diff
@@ -92,7 +92,7 @@ impl BcConnection {
     }
 
     pub fn subscribe(&self, msg_id: u32) -> Result<BcSubscription> {
-        let (tx, rx) = channel();
+        let (tx, rx) = unbounded::<Bc>();
```
Comment (QuantumEntangledAndy) [inline crates/core/src/bc_protocol/connection.rs:95]: Is this not working implictly?

```suggestion
        let (tx, rx) = unbounded();
```

File: crates/core/src/bc_protocol/errors.rs [hunk 5]
```diff
@@ -46,6 +46,10 @@ pub enum Error {
     #[error(display = "Timeout")]
     Timeout,
 
+    /// Raised when the RecvTimeoutError is reach
+    #[error(display = "Chanel Timeout")]
+    CrossbeamRecvTimeout(#[error(source)] crossbeam::RecvTimeoutError),
```
Comment (QuantumEntangledAndy) [inline crates/core/src/bc_protocol/errors.rs:51]:
```suggestion
    CrossbeamRecvTimeout(#[error(source)] crossbeam_channel::RecvTimeoutError),
```

**Comment by QuantumEntangledAndy** [comment]
Also when you send with a crossbeam chamnel to multiple receivers you do know that only one receiver will get it right?

It's sort of just  a queue of items and who every pulls it out of the channel first will get it

**Comment by entropin** [comment]
Yes, am planning to build a broadcaster  like discussed here:

https://github.com/crossbeam-rs/crossbeam-channel/issues/24

And it seems like crossbeam is more flexible, but i could be wrong

**Comment by entropin** [comment]
So what I tried and failed miserably with yesterday, was to add a Vec or a HashMap to the camera, containing Option(fn).
So that I can subscribe  bcMedia.subscribe(callback) and send a clone of all bc_media to all functions subscribed.

But rust did not let me do that across threads.  

Do you see a way of doing this without creating multiple channels and senders?

**Comment by QuantumEntangledAndy** [comment]
Hmm I've never tried registering callbacks. I've only ever needed callbacks that I can make a trait for.

What about

```rust
Arc<dyn FnMut(i32)>)
```
Although I am not sure it will work the way I think

**Comment by QuantumEntangledAndy** [comment]
I'm not sure if its helpful but maybe also check out [scoped callbacks](https://docs.rs/scoped-callback/0.2.0/scoped_callback/index.html)

**Comment by QuantumEntangledAndy** [comment]
Closures have to be handled carefully you need to ensure that everything that goes into it has to have the right lifetime to last as long as need.

I usual try a different design. Why do you want to use a closure like this? Is there another way to do it?

**Comment by entropin** [comment]
I'll try some more and come back to you with a example this is not working :D

But yes. It still feels like am trying to bend Rust due to having a object oriented / static thinking :D

**Comment by entropin** [comment]
Ok, so the problem itself is solved :D I was thinking about it all wrong.. I just needed to do a custom impl of StreamOutput's write function, collect the BcMedia there and distribute them from there, the BcMedia was easy to clone, and I can then send them both to a video processor/filter or video server like the RTSP mod, and still get stills and issue commands.

Am basically, just tetsing ideas to learn Rust, I just chose a bad one  :D But the wage goal that I use to learn, is to implement a custom web client with video preview and actions.

**Comment by QuantumEntangledAndy** [comment]
Yeah that's why it's a trait so you can implement it anyway you want. Kinda like a callback object.

**Comment by entropin** [comment]
Yes, so I'll attack it from  a different direction :) I want to support different video filters and AI prosessing of the video but still broadcast the altered stream over rtsp. 

So I'll try to find a nice way of init the rtps mod from my new "filer mod" but giving it a channel to that receives BcMedia from instead of creating it's own subscription.

**Comment by QuantumEntangledAndy** [comment]
If you only need read access to the BcMedia consider sharing it as an Arc or something so you don't need to clone it at all saving you some memory

**Comment by QuantumEntangledAndy** [comment]
Closing as it seems you found a different way to handle this. Hope it went well for you

**Pull request closed without merging**